akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -441,3 +445,102 @@ private[akka] final class ReducerState[T, R](val collector: java.util.stream.Col
   def finish(): R = collector.finisher().apply(reduced)
 }
 
+/**
+ * INTERNAL API
+ */
+final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]], zeroMat: () ⇒ M) extends GraphStageWithMaterializedValue[SinkShape[T], Future[M]] {
+  val in = Inlet[T]("lazySink.in")
+  override def initialAttributes = DefaultAttributes.lazySink
+  override val shape: SinkShape[T] = SinkShape.of(in)
+
+  override def toString: String = "LazySink"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(stoppingDecider)
+
+    val promise = Promise[M]()
+    val stageLogic = new GraphStageLogic(shape) with InHandler {
+      override def preStart(): Unit = pull(in)
+
+      override def onPush(): Unit = {
+        try {
+          val element = grab(in)
+          sinkFactory(element).onComplete {
+            case Success(sink) ⇒ initInternalSink(sink, element)
```

Comment:
I don't think so, since the "same thread" refers to the one who will complete the Future, not the thread of the stage here. You need to use AsyncCallback.invoke (but keep the sameThreadExecutionContext for efficiency, since `.invoke` is just a simple actor send, no need for a round-trip to the thread-pool just to execute that).